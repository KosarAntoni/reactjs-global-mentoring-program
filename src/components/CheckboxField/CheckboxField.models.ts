import { InputHTMLAttributes } from 'react'

export interface CheckboxFieldProps extends InputHTMLAttributes<HTMLInputElement> {
  label: string
  id: string
  className?: string
}
