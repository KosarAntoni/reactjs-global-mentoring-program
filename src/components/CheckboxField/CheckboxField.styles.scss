.checkbox-field {
  display: inline-flex;
  flex-direction: row;
  align-items: center;

  &__custom-checkbox {
    position: relative;

    width: 1.6rem;
    height: 1.6rem;
    margin-right: 0.6rem;

    background-color: var(--white);
    border-radius: 4px;

    &::after {
      content: "";

      position: absolute;
      top: 20%;
      left: 35%;
      transform: rotate(45deg);

      width: 0.2rem;
      height: 0.6rem;

      opacity: 0;
      border: solid var(--white);
      border-width: 0 2px 2px 0;
    }
  }

  &__label {
    font-size: 2rem;
    color: var(--primary-text);
  }
}

.checkbox-field__input {
  position: absolute;

  opacity: 0;

  &:checked ~ .checkbox-field__custom-checkbox {
    background-color: var(--pastele-red);

    &::after {
      opacity: 1;
    }
  }
}
