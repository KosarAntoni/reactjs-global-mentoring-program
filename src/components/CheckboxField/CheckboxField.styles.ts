import styled from '@emotion/styled'

export const StyledCheckboxField = styled.label`
  display: inline-flex;
  flex-direction: row;
  align-items: center;

  & input {
    position: absolute;

    opacity: 0;

    &:checked ~ .custom-checkbox {
      background-color: var(--pastele-red);

      &::after {
        opacity: 1;
      }
    }
  }

  & .custom-checkbox {
    position: relative;

    width: 1.6rem;
    height: 1.6rem;
    margin-right: 0.6rem;

    background-color: var(--white);
    border-radius: 4px;

    &::after {
      content: '';

      position: absolute;
      top: 20%;
      left: 35%;
      transform: rotate(45deg);

      width: 0.2rem;
      height: 0.6rem;

      opacity: 0;
      border: solid var(--white);
      border-width: 0 2px 2px 0;
    }
  }

  & .label {
    font-size: 2rem;
    color: var(--primary-text);
  }
`
