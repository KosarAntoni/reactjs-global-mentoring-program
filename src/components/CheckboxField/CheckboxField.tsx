import React, { FC } from 'react'
import classNames from 'classnames'

import { CheckboxFieldProps } from './CheckboxField.models'

const CheckboxField: FC<CheckboxFieldProps> = ({ className, label, id, ...props }) => {
  return (
    <label className={classNames('checkbox-field', className)} htmlFor={id}>
      <input type='checkbox' {...{ id, ...props }} className='checkbox-field__input'/>
      <span className='checkbox-field__custom-checkbox' />

      <span className='checkbox-field__label'>
        {label}
      </span>
    </label>
  )
}

export default CheckboxField
