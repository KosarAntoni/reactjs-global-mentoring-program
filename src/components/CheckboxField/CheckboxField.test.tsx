import React from 'react'
import { fireEvent, render, screen } from '@testing-library/react'

import CheckboxField from './CheckboxField'

import '@testing-library/jest-dom'

test('the CheckboxField component renders correctly with required and optional props', () => {
  render(<CheckboxField className='test-class' id='test' label='test label'/>)

  const node = screen.getByTestId('checkbox-node')

  // the component has been rendered
  expect(node).toBeInTheDocument()

  // generating a snapshot, not for testing purposes but only to see what changes in the HTML structure during CR if the developer updates the snapshot
  expect(node).toMatchSnapshot()

  // after click on label input is checked
  const inputNode = node.firstChild
  expect(inputNode).not.toBeChecked()
  fireEvent.click(node)
  expect(inputNode).toBeChecked()

  // has correct label text
  expect(screen.getByText('test label')).toBeInTheDocument()

  // the comonent has correct classes
  expect(node).toHaveClass('checkbox-field test-class')
})
