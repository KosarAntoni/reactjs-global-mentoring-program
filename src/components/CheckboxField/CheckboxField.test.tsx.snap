// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the CheckboxField component renders correctly with required and optional props 1`] = `
<label
  class="checkbox-field test-class"
  data-testid="checkbox-node"
  for="test"
>
  <input
    class="checkbox-field__input"
    id="test"
    type="checkbox"
  />
  <span
    class="checkbox-field__custom-checkbox"
  />
  <span
    class="checkbox-field__label"
  >
    test label
  </span>
</label>
`;
